pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/ProxyService.java

```diff
@@ -0,0 +1,243 @@
+import org.apache.pulsar.client.api.ClientConfiguration;
+import org.apache.pulsar.client.impl.ConnectionPool;
+import org.apache.pulsar.client.impl.PulsarClientImpl;
+import org.apache.pulsar.zookeeper.LocalZooKeeperConnectionService;
+import org.apache.pulsar.zookeeper.ZooKeeperClientFactory;
+import org.apache.pulsar.zookeeper.ZooKeeperSessionWatcher.ShutdownService;
+import org.apache.pulsar.zookeeper.ZookeeperClientFactoryImpl;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.netty.bootstrap.ServerBootstrap;
+import io.netty.buffer.PooledByteBufAllocator;
+import io.netty.channel.AdaptiveRecvByteBufAllocator;
+import io.netty.channel.ChannelOption;
+import io.netty.channel.EventLoopGroup;
+import io.netty.channel.epoll.EpollChannelOption;
+import io.netty.channel.epoll.EpollEventLoopGroup;
+import io.netty.channel.epoll.EpollMode;
+import io.netty.channel.epoll.EpollServerSocketChannel;
+import io.netty.channel.nio.NioEventLoopGroup;
+import io.netty.channel.socket.nio.NioServerSocketChannel;
+import io.netty.util.concurrent.DefaultThreadFactory;
+
+/**
+ * Pulsar proxy service
+ */
+public class ProxyService implements Closeable {
+
+    private final ProxyConfiguration proxyConfig;
+    private final String serviceUrl;
+    private final String serviceUrlTls;
+    private ConfigurationCacheService configurationCacheService;
+    private AuthenticationService authenticationService;
+    private AuthorizationManager authorizationManager;
+    private ZooKeeperClientFactory zkClientFactory = null;
+
+    private final EventLoopGroup acceptorGroup;
+    private final EventLoopGroup workerGroup;
+    private final DefaultThreadFactory acceptorThreadFactory = new DefaultThreadFactory("pulsar-discovery-acceptor");
+    private final DefaultThreadFactory workersThreadFactory = new DefaultThreadFactory("pulsar-discovery-io");
+
+    // ConnectionPool is used by the proxy to issue lookup requests
+    private final PulsarClientImpl client;
+
+    private final Authentication clientAuthentication;
+
+    private BrokerDiscoveryProvider discoveryProvider;
+
+    private LocalZooKeeperConnectionService localZooKeeperConnectionService;
+
+    private static final int numThreads = Runtime.getRuntime().availableProcessors();
+
+    public ProxyService(ProxyConfiguration proxyConfig) throws IOException {
+        checkNotNull(proxyConfig);
+        this.proxyConfig = proxyConfig;
+
+        String hostname;
+        try {
+            hostname = InetAddress.getLocalHost().getHostName();
+        } catch (UnknownHostException e) {
+            throw new RuntimeException(e);
+        }
+        this.serviceUrl = String.format("pulsar://%s:%d/", hostname, proxyConfig.getServicePort());
+        this.serviceUrlTls = String.format("pulsar://%s:%d/", hostname, proxyConfig.getServicePortTls());
+
+        EventLoopGroup acceptorEventLoop, workersEventLoop;
+        if (SystemUtils.IS_OS_LINUX) {
```

Comment:
I didn't know about `Epoll.isAvailable()`, probably was introduced later in Netty. Then we can just use that check, no need to verify we are on Linux.